opt: race check failure: column group contains non-equivalent columns

```
    --- FAIL: TestExecBuild/5node-dist-opt (133.81s)
        --- FAIL: TestExecBuild/5node-dist-opt/distsql_interleaved_join (21.38s)
            logic.go:2367: 
                
                testdata/distsql_interleaved_join:444: SELECT url FROM[EXPLAIN (DISTSQL)
                  SELECT * FROM grandchild1
                  JOIN child1 USING (pid1, cid1)
                  JOIN parent1 USING (pid1)
                ORDER BY pid1
                ]
                expected success, but found
                (XX000) internal error: ordering column group (5,8) contains non-equivalent columns (op inner-join)
                sql/opt/ordering/ordering.go:310: in checkRequired()
                DETAIL: stack trace:
                github.com/cockroachdb/cockroach/pkg/sql/opt/ordering/ordering.go:310: in checkRequired()
                github.com/cockroachdb/cockroach/pkg/sql/opt/ordering/ordering.go:47: in BuildChildRequired()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/physical_props.go:58: in BuildChildPhysicalProps()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:487: in optimizeGroupMember()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
                github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:209: in Optimize()
                github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:438: in buildExecMemo()
                github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:161: in makeOptimizerPlan()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:987: in makeExecPlan()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:874: in dispatchToExecutionEngine()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459: in execStmtInOpenState()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101: in execStmt()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1189: in run()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:436: in ServeConn()
                github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:577: in func1()
```

--- a/pkg/sql/opt/ordering/lookup_join.go
+++ b/pkg/sql/opt/ordering/lookup_join.go
@@ -38,7 +38,16 @@ func lookupOrIndexJoinBuildChildReqOrdering(
 
 	// We may need to remove ordering columns that are not output by the input
 	// expression.
-	return projectOrderingToInput(parent.Child(0).(memo.RelExpr), required)
+	child := parent.Child(0).(memo.RelExpr)
+	res := projectOrderingToInput(child, required)
+	// It is in principle possible that the lookup join has an ON condition that
+	// forces an equality on two columns in the input. In this case we need to
+	// trim the column groups to keep the ordering valid w.r.t the child FDs
+	// (similar to Select).
+	//
+	// This case indicates that we didn't do a good job pushing down equalities
+	// (see #36219), but it should be handled correctly here nevertheless.
+	return trimColumnGroups(&res, &child.Relational().FuncDeps)
 }
 
 func indexJoinBuildProvided(expr memo.RelExpr, required *physical.OrderingChoice) opt.Ordering {

--- a/pkg/sql/opt/testutils/opttester/opt_tester.go
+++ b/pkg/sql/opt/testutils/opttester/opt_tester.go
@@ -241,7 +241,7 @@ func (ot *OptTester) RunCommand(tb testing.TB, d *datadriven.TestData) string {
 	// Allow testcases to override the flags.
 	for _, a := range d.CmdArgs {
 		if err := ot.Flags.Set(a); err != nil {
-			d.Fatalf(tb, "%s", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 	}
 
@@ -264,7 +264,7 @@ func (ot *OptTester) RunCommand(tb testing.TB, d *datadriven.TestData) string {
 		}
 		s, err := testCatalog.ExecuteDDL(d.Input)
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		return s
 
@@ -297,7 +297,7 @@ func (ot *OptTester) RunCommand(tb testing.TB, d *datadriven.TestData) string {
 	case "opt":
 		e, err := ot.Optimize()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		ot.postProcess(tb, d, e)
 		return memo.FormatExpr(e, ot.Flags.ExprFormat)
@@ -305,35 +305,35 @@ func (ot *OptTester) RunCommand(tb testing.TB, d *datadriven.TestData) string {
 	case "optsteps":
 		result, err := ot.OptSteps()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		return result
 
 	case "exploretrace":
 		result, err := ot.ExploreTrace()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		return result
 
 	case "rulestats":
 		result, err := ot.RuleStats()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		return result
 
 	case "memo":
 		result, err := ot.Memo()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		return result
 
 	case "expr":
 		e, err := ot.Expr()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		ot.postProcess(tb, d, e)
 		return memo.FormatExpr(e, ot.Flags.ExprFormat)
@@ -341,7 +341,7 @@ func (ot *OptTester) RunCommand(tb testing.TB, d *datadriven.TestData) string {
 	case "exprnorm":
 		e, err := ot.ExprNorm()
 		if err != nil {
-			d.Fatalf(tb, "%v", err)
+			d.Fatalf(tb, "%+v", err)
 		}
 		ot.postProcess(tb, d, e)
 		return memo.FormatExpr(e, ot.Flags.ExprFormat)

--- a/pkg/sql/opt/xform/testdata/physprops/ordering
+++ b/pkg/sql/opt/xform/testdata/physprops/ordering
@@ -1552,3 +1552,23 @@ sort
            │         └── interesting orderings: (+10) (+6,+7,+10) (+8,+9,+10)
            └── filters
                 └── b = c [type=bool, outer=(2,3), fd=(2)==(3), (3)==(2)]
+
+
+# Regression test for #36219: lookup join with ON condition that imposes an
+# equality on two input columns (which isn't pushed down).
+opt disable=(PushFilterIntoJoinLeftAndRight,PushFilterIntoJoinLeft,PushFilterIntoJoinRight,MapFilterIntoJoinLeft,MapFilterIntoJoinRight)
+SELECT * FROM abc JOIN xyz ON a=x AND x=z ORDER BY z
+----
+inner-join (merge)
+ ├── columns: a:1(int!null) b:2(int!null) c:3(int!null) x:4(int!null) y:5(int!null) z:6(int!null)
+ ├── left ordering: +1
+ ├── right ordering: +4
+ ├── ordering: +(1|4|6)
+ ├── scan abc
+ │    ├── columns: a:1(int!null) b:2(int!null) c:3(int!null)
+ │    └── ordering: +1
+ ├── scan xyz
+ │    ├── columns: x:4(int!null) y:5(int!null) z:6(int!null)
+ │    └── ordering: +4
+ └── filters
+      └── x = z [type=bool]

